src/classes/AFFL_Affiliations_TDTM.cls

```diff
@@ -53,34 +48,40 @@ public class AFFL_Affiliations_TDTM extends TDTM_Runnable {
-            
-            // we used to prevent any further recursive call with contacts and affiliations,
-            // but it turns out in some scenarios, before the AfterInsert of a contact calls us,
-            // another trigger handler may update the contact, causing the AfterUpdate to come thru.
-            // if we set the flag to stop recursion, then we won't process the AfterInsert, where
-            // we create the initial affiliation. thus we no longer set a recursion prevention flag,
-            // and we only use doNotRunForContacts where we actually have a use case to prevent recursion.
-            // as for the Affiliations and Account recursion flags, we have no current known issues with
-            // their recursion prevention, so we left that as is.
-            //
-            // hasRunForContacts = true;
-            
-            return wrapper;
-        } else if (objResult.getsObjectType() == npe5__Affiliation__c.sObjectType && !hasRunForAffiliations) {
-            DmlWrapper wrapper = runAfflTriggerHandler(newlist, oldlist, triggerAction, objResult);
-            hasRunForAffiliations = true;
-            return wrapper;
-        } else if (objResult.getsObjectType() == Account.sObjectType && !hasRunForAccounts) {
-            DmlWrapper wrapper = runAccountTriggerHandler(newlist, oldlist, triggerAction, objResult);
-            hasRunForAccounts = true;
-            return wrapper;
+        if (objResult.getsObjectType() == Contact.sObjectType && !TDTM_ProcessControl.checkRecursionState(TDTM_ProcessControl.flag.AFFL_doNotRunForContacts)) {
+            wrapper = runContactTriggerHandler(newlist, oldlist, triggerAction, objResult);
+
+        } else if (objResult.getsObjectType() == npe5__Affiliation__c.sObjectType && !TDTM_ProcessControl.checkRecursionState(TDTM_ProcessControl.flag.AFFL_hasRunForAffiliations)) {
+            TDTM_ProcessControl.setRecursionFlag(TDTM_ProcessControl.flag.AFFL_hasRunForAffiliations, true);
+            wrapper = runAfflTriggerHandler(newlist, oldlist, triggerAction, objResult);
+
+        } else if (objResult.getsObjectType() == Account.sObjectType && !TDTM_ProcessControl.checkRecursionState(TDTM_ProcessControl.flag.AFFL_hasRunForAccounts)) {
+            TDTM_ProcessControl.setRecursionFlag(TDTM_ProcessControl.flag.AFFL_hasRunForAccounts, true);
+            wrapper = runAccountTriggerHandler(newlist, oldlist, triggerAction, objResult);
+
         } else {
             return null;
         }
+
+        // Process the DML
+        TDTM_TriggerHandler.processDML(wrapper);
+
+        // Decrement the recursion counters based on related processing logic
+        if (objResult.getsObjectType() == npe5__Affiliation__c.sObjectType) {
+            TDTM_ProcessControl.setRecursionFlag(TDTM_ProcessControl.flag.AFFL_hasRunForAffiliations, false);
+            if (TDTM_ProcessControl.checkRecursionState(TDTM_ProcessControl.flag.AFFL_hasRunForAccounts)) {
```

Comment:
This is effectively the original logic substituting our flags. Though what I remember seeing with the logic here is that after DML, if the Affiliation trigger is running on the Affiliation object it clears the flag for hasRunForAffiliations, and if the Affiliations trigger on Account had previously run it clears the doNotRunForContacts flag because that flag is set in the Affiliations-Account logic to prevent the trigger from running on the Contact when the Affiliation is created/updated. This is one of those scenarios where a single TDTM trigger class is called by triggers on the Account, Contact and Affiliation objects.